agent/src/main/scala/za/co/absa/atum/agent/model/MeasurementBuilder.scala

```diff
@@ -16,19 +16,40 @@
 
 package za.co.absa.atum.agent.model
 
+import za.co.absa.atum.agent.exception.AtumAgentException.MeasurementException
 import za.co.absa.atum.model.dto.{MeasureDTO, MeasureResultDTO, MeasurementDTO}
-import za.co.absa.atum.model.dto.MeasureResultDTO.TypedValue
 
 /**
  * This object provides a functionality to convert a measurement to its DTO representation.
  */
 private [agent] object MeasurementBuilder {
 
-  private [agent] def buildMeasurementDTO(measurement: Measurement): MeasurementDTO = {
-    val measureName = measurement.measure.measureName
-    val measureDTO = MeasureDTO(measureName, measurement.measure.controlColumns)
-    val measureResultDTO = MeasureResultDTO(TypedValue(measurement.resultValue.toString, measurement.resultType))
+  private[agent] def validateMeasureUniqueness(measurements: Set[Measurement]): Unit = {
+    val allMeasures = measurements.toSeq.map(_.measure)
 
+    val originalMeasuresCnt = allMeasures.size
+    val uniqueMeasureColumnCombinationCnt = allMeasures.map(m =>
+      (m.measureName, m.measuredColumns)  // there can't be two same measures defined on the same column(s)
+    ).distinct.size
+
+    if (originalMeasuresCnt != uniqueMeasureColumnCombinationCnt)
+      throw MeasurementException(
+        s"Measure and measuredColumn combinations must be unique, i.e. they cannot repeat! Got: $allMeasures"
+      )
+  }
+
+  private[agent] def buildMeasurementsDTO(measurements: Set[Measurement]): Set[MeasurementDTO] = {
+    validateMeasureUniqueness(measurements)
```

Comment:
Why the validation? If not set, duplicates will be eliminated. So what?

Reply:
Depends what you mean by duplicates. Some scenarios will be eliminated, some not, I want to cover it all.

These duplicates would be eliminated by the nature of set being... set :)

```scala
      Measurement(RecordCount(), MeasureResult("1", ResultValueType.Long)),
      Measurement(SumOfValuesOfColumn("col1"), MeasureResult(BigDecimal(1.2))),
      Measurement(SumOfValuesOfColumn("col1"), MeasureResult(BigDecimal(1.2))),
      Measurement(DistinctRecordCount(Seq("col2")), MeasureResult(3L))
```
^ this is actually 3 unique records

--

but, this is actually 4 records:

```
      Measurement(RecordCount(), MeasureResult("1", ResultValueType.Long)),
      Measurement(SumOfValuesOfColumn("col1"), MeasureResult(BigDecimal(1.2))),
      Measurement(SumOfValuesOfColumn("col1"), MeasureResult(BigDecimal(1.4))),
      Measurement(DistinctRecordCount(Seq("col2")), MeasureResult(3L))
```

so there is a 'conflict' that the SET can't fix - second and third measurements are supposed to be the same - same Atum Context captured the same Measure on the same column but with different results. So I need to 'dig deeper' into the Set of Measurement objects - I need to check all Measures and their columns, and I need to make sure that only unique combination of these is further used

after all, there are unit tests testing (validity and non-valid cases) all this